edb/pgsql/delta.py

```diff
@@ -888,7 +906,158 @@ def compile_edgeql_function(self, func: s_funcs.Function, schema, context):
+        set_returning = func_return_typemod is ql_ft.TypeModifier.SetOfType
+        ov_params = overloads[0].get_params(schema).objects(schema)
+        my_params = func.get_params(schema).objects(schema)
+        my_pt = tuple(p.get_type(schema) for p in my_params)
+        ov_pt = (p.get_type(schema) for p in ov_params)
+        ov_param_idx = -1
+        for i, (my_t, ov_t) in enumerate(zip(my_pt, ov_pt)):
+            if my_t != ov_t:
+                if ov_param_idx != -1:
+                    raise AssertionError(
+                        "unexpected overload on multiple parameters "
+                        "in object-taking function"
+                    )
+                ov_param_idx = i
+
+        if (
+            ov_param_idx == -1
+            or not my_pt[ov_param_idx].is_object_type()
+        ):
+            raise AssertionError(
+                "could not find generic parameter in "
+                "overloaded function"
+            )
+
+        param_name = my_params[ov_param_idx].get_parameter_name(schema)
+        type_param_name = f'__{param_name}__type'
+        cases = {}
+        all_overloads = list(overloads)
+        if not isinstance(self, DeleteFunction):
+            all_overloads.append(func)
+        for overload in all_overloads:
+            ov_p = tuple(overload.get_params(schema).objects(schema))
+            ov_p_t = ov_p[ov_param_idx].get_type(schema)
+            ov_body = self.compile_edgeql_function_body(
+                overload, schema, context)
+
+            if set_returning:
+                case = (
+                    f"(SELECT * FROM ({ov_body}) AS q "
+                    f"WHERE ancestor = {ql(str(ov_p_t.id))})"
+                )
+            else:
+                case = (
+                    f"WHEN ancestor = {ql(str(ov_p_t.id))} "
+                    f"THEN \n({ov_body})"
+                )
+
+            cases[ov_p_t] = case
+
+        impl_ids = ', '.join(f'{ql(str(t.id))}::uuid' for t in cases)
+        branches = list(cases.values())
+
+        # N.B: edgedb.raise and coalesce are used below instead of
+        #      raise_on_null, because the latter somehow results in a
+        #      significantly more complex query plan.
+        matching_impl = f"""
+            coalesce(
+                (
+                    SELECT
+                        ancestor
+                    FROM
+                        (SELECT
+                            {qi(type_param_name)} AS ancestor,
+                            -1 AS index
+                        UNION ALL
+                        SELECT
+                            target AS ancestor,
+                            index
+                        FROM
+                            edgedb."_SchemaObjectType__ancestors"
+                            WHERE source = {qi(type_param_name)}
+                        ) a
+                    WHERE ancestor IN ({impl_ids})
+                    ORDER BY index
```

Comment:
Would it be worth it to try to put an index on (source, index)? Probably not, because the sets probably aren't very big?

Reply:
There is a unique `(source, target)` index on all links, and that is also necessary to enforce proper distinctness.